pallets/liquid-staking-two/src/lib.rs

```diff
@@ -0,0 +1,820 @@
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Liquid staking pallet
+//!
+//! ## Overview
+//!
+//! This pallet manages the NPoS operations for relay chain asset.
+
+// 1. update parachain era before relaychain era updated, a few blocks in advance,
+// 2. calculate parachain status, whether bond/unbond/rebond,
+// 3. invoke relaychain staking methods, bond/unbond/rebond,
+// 4. invoke parachain method record relaychain resonse, whether bond/unbond successed or failed,
+// 5. waiting relaychain era updated, get the reward amount.
+// 6. record in parachain with blocknumber when real relaychain era updated.
+// 7. record reward on parachain. update parachain exchange rate.
+// 8. if step-4 successed, mint/deposit xToken out according to current exchangerate(after record reward),
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+    pallet_prelude::*, traits::SortedMembers, transactional, BoundedVec, PalletId,
+};
+use frame_system::pallet_prelude::*;
+use orml_traits::XcmTransfer;
+use sp_runtime::{traits::AccountIdConversion, ArithmeticError, FixedPointNumber, RuntimeDebug};
+use sp_std::convert::TryInto;
+use sp_std::prelude::*;
+use xcm::v0::{Junction, MultiLocation, NetworkId};
+
+use orml_traits::{MultiCurrency, MultiCurrencyExtended};
+
+pub use pallet::*;
+use primitives::{
+    Amount, Balance, BlockNumber, CurrencyId, EraIndex, ExchangeRateProvider,
+    LiquidStakingProtocol, Rate, Ratio,
+};
+
+pub const MAX_UNSTAKE_CHUNKS: usize = 5;
+
+#[derive(Copy, Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug)]
+pub enum StakingOperationType {
+    Bond,
+    Unbond,
+    Rebond,
+    Matching,
+    TransferToRelaychain,
+    RecordReward,
+    RecordSlash,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug)]
+pub enum ResponseStatus {
+    Ready,
+    Processing,
+    Successed,
+    Failed,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug)]
+pub struct Operation<BlockNumber> {
+    amount: Balance,
+    block_number: BlockNumber,
+    status: ResponseStatus,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Default, Encode, Decode, RuntimeDebug)]
+pub struct MatchingPoolBuffer {
+    total_unstake_amount: Balance,
+    total_stake_amount: Balance,
+    operation_type: Option<StakingOperationType>,
+}
+
+/// The single user's stake/unsatke amount in each era
+#[derive(Copy, Clone, Eq, PartialEq, Default, Encode, Decode, RuntimeDebug)]
+pub struct MatchingUserBuffer {
+    /// The token amount that user unstake during this era, will be calculated
+    /// by exchangerate and xToken amount
+    total_unstake_amount: Balance,
```

Comment:
compacted balance should be preferred

Reply:
I guess compacted balance should only be used in extrinsic signature.